code/datums/components/quality.dm

```diff
@@ -0,0 +1,78 @@
+#define UNUSABLE_QUALITY 0.7
+#define SHODDY_QUALITY 0.8
+#define POOR_QUALITY 0.9
+#define NORMAL_QUALITY 1
+#define GOOD_QUALITY 1.05
+#define FINE_QUALITY 1.1
+#define SUPERIOR_QUALITY 1.2
+#define EXCEPTIONAL_QUALITY 1.3
+#define MASTERFUL_QUALITY 1.5
+#define ARTIFACT_QUALITY 1.75 // big upgrade coz it is so rare
+
+datum/component/quality
+	dupe_mode = COMPONENT_DUPE_UNIQUE_PASSARGS
+
+	var/list/quality_levels = list(UNUSABLE_QUALITY,SHODDY_QUALITY,POOR_QUALITY,NORMAL_QUALITY,GOOD_QUALITY,FINE_QUALITY,SUPERIOR_QUALITY,EXCEPTIONAL_QUALITY,MASTERFUL_QUALITY,ARTIFACT_QUALITY )
+	var/quality_level
+	var/old_name
+
+	var/obj/item/parent_item
+
+	var/mob/living/carbon/human/creator
+	var/datum/mind/creator_mind
+	var/skill_level
+
+	var/list/quality_list = list(0,1,1,2,2,3)
+
+/datum/component/quality/Initialize(mob/living/carbon/human/_creator,datum/skill/_skill,modifier)
+	if(!isitem(parent) || !_creator)
+		return COMPONENT_INCOMPATIBLE
+	creator = _creator
+	skill_level =  creator?.mind.get_skill_level(_skill)
+	creator_mind =  creator?.mind
+	parent_item = parent
+	old_name = parent_item.name
+	generate_quality(skill_level,modifier)
+	apply_quality()
+
+
+/datum/component/quality/proc/generate_quality(skill_level,modifier = 0 )
+	quality_level = clamp(pick(quality_list) + skill_level + modifier,0,9)
```

Comment:
Add some comments about the math here, I don't really want to have to look up all these values (Max level, max quality level etc)

Reply:
Also I very much dislike this method of determining quality, you can essentialy never get a bad item when you're skill and when you're at level 6 you basically make an artifact once ever 6 items which is way too high. IMO artifacts should be locked behind some extra work like having a status effect for killing a megafauna (Which makes it more of a challenge to get one)